### PR TITLE
planner/core: move all extra columns to the end in ExpandVirtualColumn

### LINKED ISSUE #58475: select failed for partition table with generated column with error `needPartitionHandle != ret, tp(1)`
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```sql
drop table tp;

CREATE TABLE tp (
  id int, c1 json, c2 json GENERATED ALWAYS AS (c1) VIRTUAL, KEY `idx_31` (id)
) PARTITION BY RANGE (id)
(PARTITION `p0` VALUES LESS THAN (2),
 PARTITION `p1` VALUES LESS THAN (4532022));

INSERT INTO tp (id, c1) VALUES (0, "[1]");

select id from tp where json_contains(c2, "1") group by id having id in (0);
```

<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)

```sql
+------+                                                                                                                                                                                                         
| id   |                                                                                                                                                                                                         
+------+                                                                                                                                                                                                         
|    0 |                                                                                                                                                                                                         
+------+ 
```

### 3. What did you see instead (Required)

```
ERROR 1105 (HY000): Internal error, needPartitionHandle != ret, tp(1)
```

### 4. What is your TiDB version? (Required)

<!-- Paste the output of SELECT tidb_version() -->

### PR DESCRIPTION
**Pull request by joechenrh**
### What problem does this PR solve?

Issue Number: close #58475, close https://github.com/pingcap/tidb/issues/56013

Problem Summary:

### What changed and how does it work?

As title. Previously we only move column with id `ExtraHandleID` to the end, but we forgot some other extra columns.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/planner/core/integration_test.go [hunk 1]
```diff
@@ -2318,3 +2318,23 @@ func TestNestedVirtualGeneratedColumnUpdate(t *testing.T) {
 	tk.MustExec("UPDATE test1 SET col7 = '{\"col10\":\"DDDDD\",\"col9\":[\"abcdefg\"]}';\n")
 	tk.MustExec("DELETE FROM test1 WHERE col1 < 0;\n")
 }
+
+// Issue 58475
+func TestGeneratedColumnWithPartition(t *testing.T) {
```
Comment (hawkingrei) [inline pkg/planner/core/integration_test.go:2323]: move this test case into ```pkg/planner/core/casetest/partition```.

File: pkg/planner/core/physical_plans.go [hunk 2]
```diff
@@ -1059,16 +1059,28 @@ func (ts *PhysicalTableScan) ResolveCorrelatedColumns() ([]*ranger.Range, error)
...
+	copyColumn := make([]*model.ColumnInfo, 0, len(columns))
+	copyColumn = append(copyColumn, columns...)
+
+	oldNumColumns := len(schema.Columns)
+	numExtraColumns := 0
+	for i := oldNumColumns - 1; i >= 0; i-- {
+		cid := schema.Columns[i].ID
+		// All columns with negative cid should be placed at last.
+		if cid < 0 {
+			break
```
Comment (dash12653) [inline pkg/planner/core/physical_plans.go:1071]: Can we set `break `to `continue`?  Maybe issue #56013 could be resolved at the same time.

Comment (Defined2014) [reply pkg/planner/core/physical_plans.go:1071]: I think it should be `cid > 0`

File: pkg/planner/core/physical_plans.go [hunk 3]
```diff
@@ -1059,16 +1059,28 @@ func (ts *PhysicalTableScan) ResolveCorrelatedColumns() ([]*ranger.Range, error)
...
-		copyColumn = copyColumn[:len(copyColumn)-1]
+	copyColumn := make([]*model.ColumnInfo, 0, len(columns))
+	copyColumn = append(copyColumn, columns...)
+
+	oldNumColumns := len(schema.Columns)
+	numExtraColumns := 0
+	for i := oldNumColumns - 1; i >= 0; i-- {
+		cid := schema.Columns[i].ID
+		// All columns with negative cid should be placed at last.
+		if cid < 0 {
```
Comment (Defined2014) [inline pkg/planner/core/physical_plans.go:1070]:
```suggestion
		if cid > 0 {
```

**Review by Defined2014 — APPROVED** [review]

**Review by hawkingrei — APPROVED** [review]

**Pull request merged**